core/src/main/java/org/infinispan/distribution/TriangleOrderManager.java

```diff
@@ -0,0 +1,187 @@
+   }
+
+   public void removeOldSegments(Collection<Integer> ownedSegments) {
+      if (ownedSegments.isEmpty()) {
+         //no segments?
+         receiverSegmentMap.clear();
+         senderSegmentMap.clear();
+      } else {
+         receiverSegmentMap.keySet().retainAll(ownedSegments);
+         senderSegmentMap.keySet().retainAll(ownedSegments);
+      }
+   }
+
+   private long getNext(int segmentId, int topologyId) {
+      return senderSegmentMap.computeIfAbsent(segmentId, integer -> new SenderSequencer(topologyId)).next(topologyId);
+   }
+
+   private boolean checkIfNext(int segmentId, int topologyId, long sequenceNumber) {
+      return receiverSegmentMap.computeIfAbsent(segmentId, integer -> new ReceiverSequencer(topologyId))
+            .isNext(topologyId, sequenceNumber);
+   }
+
+   private boolean checkAllEntries(Map<Integer, Long> sequenceNumbers, int commandTopologyId) {
+      Iterator<Map.Entry<Integer, Long>> iterator = sequenceNumbers.entrySet().iterator();
+      boolean result = true;
+      while (result && iterator.hasNext()) {
+         Map.Entry<Integer, Long> entry = iterator.next();
+         result = checkIfNext(entry.getKey(), commandTopologyId, entry.getValue());
+      }
+      return result;
+   }
+
+   private void checkTopologyId(int topologyId) {
+      if (topologyId != currentCacheTopology.getTopologyId()) {
+         throw OutdatedTopologyException.getCachedInstance();
+      }
+   }
+
+   private class SenderSequencer {
+      @GuardedBy("this")
+      private int topologyId;
+      @GuardedBy("this")
+      private long sequenceNumber = 1;
+
+      private SenderSequencer(int topologyId) {
+         setTopologyId(topologyId);
+      }
+
+      private synchronized long next(int commandTopologyId) {
+         if (topologyId == commandTopologyId) {
+            return sequenceNumber++;
+         } else if (topologyId < commandTopologyId) {
+            //update topology. this command will be the first
+            setTopologyId(commandTopologyId);
+            return sequenceNumber++;
+         } else {
+            //this topology is higher than the command topology id.
+            //another topology was installed. this command will fail with OutdatedTopologyException.
+            throw OutdatedTopologyException.getCachedInstance();
+         }
+      }
+
+      private void setTopologyId(int topologyId) {
+         this.topologyId = topologyId;
+         this.sequenceNumber = 1;
+      }
+   }
+
+   private class ReceiverSequencer {
+      @GuardedBy("this")
+      private int topologyId;
+      @GuardedBy("this")
+      private long nextSequenceNumber = 1;
+
+      private ReceiverSequencer(int topologyId) {
+         setTopologyId(topologyId);
+      }
+
+      private synchronized void deliver(int commandTopologyId, long sequenceNumber) {
+         if (topologyId == commandTopologyId && nextSequenceNumber == sequenceNumber) {
+            nextSequenceNumber++;
```

Comment:
I feel like there should be a way to trigger the next command directly here, instead of having to call `checkForReadyTasks()` after, because that method has to iterate through all the tasks in the queue and it's never going to run more than 1 task per segment.

OTOH it's probably good that `checkForReadyTasks()` can pick a command for any segment, and one segment's  commands can't monopolize a remote executor thread forever, so I'm not pushing for a change yet.